Jellyfin.Server/Program.cs

```diff
@@ -1,313 +1,335 @@
-                commandLineArgsString = string .Join(" ", 
-                    Environment.GetCommandLineArgs()
-                        .Skip(1)
-                        .Select(NormalizeCommandLineArgument)
-                    );
-            }
-
-            _logger.LogInformation("Executable: {0}", module);
-            _logger.LogInformation("Arguments: {0}", commandLineArgsString);
-
-            Process.Start(module, commandLineArgsString);
-        }
-
-        private static string NormalizeCommandLineArgument(string arg)
-        {
-            if (!arg.Contains(" ", StringComparison.OrdinalIgnoreCase))
-            {
-                return arg;
-            }
-
-            return "\"" + arg + "\"";
-        }
-    }
-}
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Security;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Emby.Drawing;
+using Emby.Drawing.Skia;
+using Emby.Server.Implementations;
+using Emby.Server.Implementations.EnvironmentInfo;
+using Emby.Server.Implementations.IO;
+using Emby.Server.Implementations.Networking;
+using MediaBrowser.Common.Configuration;
+using MediaBrowser.Common.Net;
+using MediaBrowser.Controller.Drawing;
+using MediaBrowser.Model.IO;
+using MediaBrowser.Model.Globalization;
+using MediaBrowser.Model.System;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using Serilog;
+using Serilog.AspNetCore;
+using ILogger = Microsoft.Extensions.Logging.ILogger;
+
+namespace Jellyfin.Server
+{
+    public static class Program
+    {
+        private static readonly CancellationTokenSource _tokenSource = new CancellationTokenSource();
+        private static readonly ILoggerFactory _loggerFactory = new SerilogLoggerFactory();
+        private static ILogger _logger;
+        private static bool _restartOnShutdown;
+
+        public static async Task Main(string[] args)
+        {
+            StartupOptions options = new StartupOptions(args);
+            Version version = Assembly.GetEntryAssembly().GetName().Version;
+
+            if (options.ContainsOption("-v") || options.ContainsOption("--version"))
+            {
+                Console.WriteLine(version.ToString());
+            }
+
+            ServerApplicationPaths appPaths = createApplicationPaths(options);
+            // $JELLYFIN_LOG_DIR needs to be set for the logger configuration manager
+            Environment.SetEnvironmentVariable("JELLYFIN_LOG_DIR", appPaths.LogDirectoryPath);
+            await createLogger(appPaths);
+            _logger = _loggerFactory.CreateLogger("Main");
+
+            AppDomain.CurrentDomain.UnhandledException += (sender, e)
+                => _logger.LogCritical((Exception)e.ExceptionObject, "Unhandled Exception");
+
+            // Intercept Ctrl+C and Ctrl+Break
+            Console.CancelKeyPress += (sender, e) =>
```

Comment:
Does this also catch SIGINT when sent from a kill command for example? (Or an orchestrator)
A tty might not be connected.

Reply:
```
Occurs when the System.ConsoleModifiers.Control modifier key (Ctrl) and either the System.ConsoleKey.C console key (C) or the Break key are pressed simultaneously (Ctrl+C or Ctrl+Break).
```
Doesn't look like it.

Reply:
I've done some digging and windows needs key handlers while linux needs signal handling. Is it possible to do both? If there is no way to register SIGINT, I bet that the linux implementation will catch SIGINT and emulate Ctrl+C to the application. Don't quote me on this though.

Reply:
And that seems to not be in dotnet core anyway. Otherwise we could do something with unsafe code.